Add the first global topology test

ElhadiCherifi95 opened the pull request.

ElhadiCherifi95 committed:
Add tests on all topology + correcting fizz buzz logic

DivLoic commented on pom.xml:
```diff
@@ -23,7 +23,8 @@
         <!-- dependencies -->
-        <junit.version>5.5.1</junit.version>
+        <junit.jupiter.version>5.5.1</junit.jupiter.version>
+        <junit.version>4.12</junit.version>
```
I'd like to avoid having two versions of JUnit. Is 4 in mandatory to `TopologyTestDriver`?

DivLoic commented on fizz-buzz-streams/src/test/resources/application.conf:
```diff
@@ -0,0 +1,6 @@
+application.id = LOCAL-BUZZ-APP
```
`TEST-BUZZ-APP`? So it's different from the local runs.

DivLoic commented on fizz-buzz-streams/src/test/resources/application.conf:
```diff
@@ -0,0 +1,6 @@
+application.id = LOCAL-BUZZ-APP
+//bootstrap servers configuration value won't be usable
```
Delete this

DivLoic commented on fizz-buzz-streams/pom.xml:
```diff
@@ -87,6 +87,12 @@
             <version>${kafka.version}</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Can we avoid including this previous version?

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+import static org.hamcrest.core.IsEqual.equalTo;
+import static org.junit.Assert.assertThat;
+
+public class FizzBuzzAllTopologySpec {
```
🙂Let's stick with Java convention and call it a Test (not a Spec) `FizzBuzzCompleteTopologyTest`

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+    @Before
+    public void setTopologyTestDriver() throws IOException, RestClientException {
+
+        final String srConfigKey = "schema.registry.url";
```
May be static and at class level

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+        properties.put(StreamsConfig.APPLICATION_ID_CONFIG, config.getString("application.id"));
+        properties.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest");
+        properties.put(StreamsConfig.CACHE_MAX_BYTES_BUFFERING_CONFIG, 0);
+
```
Too much space

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+
+        MockSchemaRegistryClient mockedRegistry = new MockSchemaRegistryClient();
+
+
```
Too much space

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+
+        TimeWindows windows = TimeWindows
+
+                .of(Duration.ofSeconds(20))
```
Latest version creates the window from the `Config` object. This can be applied here.

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    @After
+    public void closeTestDriver(){
```
Suggestion: be less specific with the name `tearDown()`, so every java fan can relate to it.

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+
+        //Given
+        final List<KeyValue<InputKey,Input>> inputValues = Arrays.asList(
+                new KeyValue<>(new InputKey("player-1"),new Input(3,Instant.parse("2020-02-14T14:26:00Z"))),
```
check space and refactor in general

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    @Test
+    public void allTopologyTestDifferentPlayersSameWindow() {
```
Suggestion:  `topologyShouldGroupEventsFromSameClient()`
BTW: let's forget the idea of _"player"_ (this is a random thing I wrote) and describe the Sapients as "client" so (`client-1`, `client-2`)

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    @Test
+    public void allTopologyTestSamePlayerDifferentWindows() {
```
Suggestion: `topologyShouldGroupEventsFromSameWindow()`

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    @Test
+    public void allTopologyTestDifferentPlayersDifferentWindows() {
```
We can get rid of this one.

DivLoic commented on fizz-buzz-streams/src/test/java/com/pubsap/eng/streams/FizzBuzzAllTopologySpec.java:
```diff
@@ -0,0 +1,192 @@
+    }
+
+    @Test
+    public void allTopologyTestSamePlayerSameWindow() {
```
Suggestion: `topologyShouldGroupEvents()`

DivLoic reviewed (changes requested):
Please, update the branch feature/add_all_topology_tests before applying the requested changes!

ElhadiCherifi95 committed:
Applying MR suggestions:
+Delete JUnit 4 dependency
+Refactoring tests names
+Load window parameters from Config

ElhadiCherifi95 committed:
Merge branch 'master' into feature/add_all_topology_tests
